osu.Game.Rulesets.Osu/Objects/Drawables/SliderInputManager.cs

```diff
@@ -0,0 +1,254 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Input;
+using osu.Framework.Input.Events;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Objects.Types;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Objects.Drawables
+{
+    public partial class SliderInputManager : Component, IRequireHighFrequencyMousePosition
+    {
+        /// <summary>
+        /// Whether the slider is currently being tracked.
+        /// </summary>
+        public bool Tracking { get; private set; }
+
+        /// <summary>
+        /// The point in time after which we can accept any key for tracking. Before this time, we may need to restrict tracking to the key used to hit the head circle.
+        ///
+        /// This is a requirement to stop the case where a player holds down one key (from before the slider) and taps the second key while maintaining full scoring (tracking) of sliders.
+        /// Visually, this special case can be seen below (time increasing from left to right):
+        ///
+        ///  Z  Z+X  Z
+        ///      o========o
+        ///
+        /// Without this logic, tracking would continue through the entire slider even though no key hold action is directly attributing to it.
+        ///
+        /// In all other cases, no special handling is required (either key being pressed is allowable as valid tracking).
+        ///
+        /// The reason for storing this as a time value (rather than a bool) is to correctly handle rewind scenarios.
+        /// </summary>
+        private double? timeToAcceptAnyKeyAfter;
+
+        /// <summary>
+        /// The actions that were pressed in the previous frame.
+        /// </summary>
+        private readonly List<OsuAction> lastPressedActions = new List<OsuAction>();
+
+        private Vector2? screenSpaceMousePosition;
+        private readonly DrawableSlider slider;
+
+        public SliderInputManager(DrawableSlider slider)
+        {
+            this.slider = slider;
+        }
+
+        /// <summary>
+        /// This component handles all input of the slider, so it should receive input no matter the position.
+        /// </summary>
+        public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => true;
+
+        protected override bool OnMouseMove(MouseMoveEvent e)
+        {
+            screenSpaceMousePosition = e.ScreenSpaceMousePosition;
+            return base.OnMouseMove(e);
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+            updateTracking(isMouseInFollowArea(Tracking));
+        }
+
+        public void PostProcessHeadJudgement(DrawableSliderHead head)
+        {
+            if (!head.Judged || !head.Result.IsHit)
+                return;
+
+            if (!isMouseInFollowArea(true))
```

Comment:
This is arguable, but I think it might make more sense to move this check to after the foreach ?

Reply:
Disagree

Reply:
Disagree even with the fact that 2nd and 4th being different is weird ? Not that it matters that much anyway, but I just felt like it was a tiny bit weird.